event_reporter/sharding/sharding.go

```diff
@@ -0,0 +1,102 @@
+package sharding
+
+import (
+	"fmt"
+	argocommon "github.com/argoproj/argo-cd/v2/common"
+	"github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	"hash/fnv"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/argoproj/argo-cd/v2/util/env"
+	log "github.com/sirupsen/logrus"
+)
+
+var osHostnameFunction = os.Hostname
+
+type DistributionFunction func(c *v1alpha1.Application) int
+type ApplicationFilterFunction func(c *v1alpha1.Application) (bool, int)
+
+type Sharding interface {
+	GetApplicationFilter(distributionFunction DistributionFunction, shard int) ApplicationFilterFunction
+	GetDistributionFunction(shardingAlgorithm string) DistributionFunction
+}
+
+type sharding struct {
+}
+
+func NewSharding() Sharding {
+	return &sharding{}
+}
+
+func (s *sharding) GetApplicationFilter(distributionFunction DistributionFunction, shard int) ApplicationFilterFunction {
+	return func(app *v1alpha1.Application) (bool, int) {
+		expectedShard := distributionFunction(app)
+		// TODO: [reporter] provide ability define label with shard number
+		return expectedShard == shard, expectedShard
+	}
+}
+
+// GetDistributionFunction returns which DistributionFunction should be used based on the passed algorithm and
+// the current datas.
+func (s *sharding) GetDistributionFunction(shardingAlgorithm string) DistributionFunction {
```

Comment:
will be here err in future ?

Reply:
i think no, it always will be some algorithm that is default